src/compiler/checker.ts

```diff
@@ -17673,6 +17673,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
         // We optimize for the common case of unioning a union type with some other type (such as `undefined`).
         if (types.length === 2 && !origin && (types[0].flags & TypeFlags.Union || types[1].flags & TypeFlags.Union)) {
+            // avoid subtype reduction when one of the types has only types that can't have subtypes
+            if (unionReduction === UnionReduction.Subtype && some(types, t => everyType(t, t => !!(t.flags & TypeFlags.Nullable)))) {

```

Comment:
I'm not sure if this is a great idea, this should be a somewhat common case though. This check could be expanded to cover more types that can't be subtypes (like symbols) but I wouldn't expect those cases to be that common so perhaps it's not worth it.